### PR TITLE
[Fix] Fix MaskFormer and Mask2Former

### PR DESCRIPTION
**Pull request by Li-Qingyun**
## Motivation

The DETR-related modules have been refactored in #8763, which causes breakings of MaskFormer and Mask2Former. The unit tests of them were deleted in #9089. This pr add the ut back, and fix the breakings. Note that this pr only fix the bugs, but not refactor the two detector. Their refactors require new prs and more time for designing scheme.

## BC-breaking

The modifications may causes breaking of the MaskFormer and Mask2Former newly supported in open-mmlab/mmsegmentation#2215 and open-mmlab/mmsegmentation#2255 of mmsegmentation v1.0.0rc2. 
(The weight in the model zoo should be converted.)

### CODE REVIEW

**Comment by Li-Qingyun** [comment]
@Czm369 This work is directed by @jshilong.

**Comment by Li-Qingyun** [comment]
## Align Inference Accuracy
.\configs\mask2former\mask2former_r50_8xb2-lsj-50e_coco-panoptic.py
.\develop\new_mask2former_r50_lsj_8x2_50e_coco-panoptic_20220326_224516-11a44721.pth (need to be converted)

```
+--------+--------+--------+--------+------------+
|        | PQ     | SQ     | RQ     | categories |
+--------+--------+--------+--------+------------+
| All    | 51.865 | 83.071 | 61.591 | 133        |
| Things | 57.737 | 84.043 | 68.129 | 80         |
| Stuff  | 43.003 | 81.604 | 51.722 | 53         |
+--------+--------+--------+--------+------------+
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.448
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.419
12/20 23:29:29 - mmengine - INFO - Iter(test) [5000/5000]  coco_panoptic/PQ: 51.8654  coco_panoptic/SQ: 83.0710  coco_panoptic/RQ: 61.5908  coco_panoptic/PQ_th: 57.7368  coco_panoptic/SQ_th: 84.0430  coco_panoptic/RQ_th: 68.1286  coco_panoptic/PQ_st: 43.0029  coco_panoptic/SQ_st: 81.6037  coco_panoptic/RQ_st: 51.7223  coco/bbox_mAP: 0.4480  coco/bbox_mAP_50: 0.6440  coco/bbox_mAP_75: 0.4800  coco/bbox_mAP_s: 0.2650  coco/bbox_mAP_m: 0.4800  coco/bbox_mAP_l: 0.6250  coco/segm_mAP: 0.4190  coco/segm_mAP_50: 0.6460  coco/segm_mAP_75: 0.4480  coco/segm_mAP_s: 0.2150  coco/segm_mAP_m: 0.4590  coco/segm_mAP_l: 0.6340
```

The checkpoint should be converted by the following script:
```Python
import json
from collections import OrderedDict

import torch

from mmengine.config import Config
from mmdet.models import build_detector
from mmdet.utils import register_all_modules
register_all_modules(init_default_scope=True)


def get_new_name(old_name: str):
    new_name = old_name

    if 'encoder.layers' in new_name:
        new_name = new_name.replace('attentions.0', 'self_attn')

    new_name = new_name.replace('ffns.0', 'ffn')

    if 'decoder.layers' in new_name:
        new_name = new_name.replace('attentions.0', 'cross_attn')
        new_name = new_name.replace('attentions.1', 'self_attn')

    return new_name


def cvt_sd(old_sd: OrderedDict):
    new_sd = OrderedDict()
    for name, param in old_sd.items():
        new_name = get_new_name(name)
        assert new_name not in new_sd
        new_sd[new_name] = param
    assert len(new_sd) == len(old_sd)
    return new_sd


if __name__ == '__main__':

    CFG_FILE = 'configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco-panoptic.py'
    OLD_CKPT_FILENAME = 'mask2former_r50_lsj_8x2_50e_coco-panoptic_20220326_224516-11a44721.pth'
    OLD_CKPT_FILEPATH = 'develop/' + OLD_CKPT_FILENAME
    NEW_CKPT_FILEPATH = 'develop/new_' + OLD_CKPT_FILENAME

    cfg = Config.fromfile(CFG_FILE)
    model_cfg = cfg.model

    detector = build_detector(model_cfg)

    refer_sd = detector.state_dict()
    old_sd = torch.load(OLD_CKPT_FILEPATH)['state_dict']

    new_sd = cvt_sd(old_sd)

    new_names = sorted(list(refer_sd.keys()))
    cvt_names = sorted(list(new_sd.keys()))
    old_names = sorted(list(old_sd.keys()))

    # we should make cvt_names --> new_names
    json.dump(new_names, open(r'./develop/new_names.json', 'w'), indent='\n')
    json.dump(cvt_names, open(r'./develop/cvt_names.json', 'w'), indent='\n')
    json.dump(old_names, open(r'./develop/old_names.json', 'w'), indent='\n')

    new_ckpt = dict(state_dict=new_sd)
    torch.save(new_ckpt, NEW_CKPT_FILEPATH)
    print(f'{NEW_CKPT_FILEPATH} has been saved!')
```

**Comment by jshilong** [comment]
please merge the refactor-detr and resolve the conflict

File: configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco-panoptic.py [hunk 1]
```diff
@@ -55,62 +55,45 @@
...
-                transformerlayers=dict(
-                    type='BaseTransformerLayer',
-                    attn_cfgs=dict(
-                        type='MultiScaleDeformableAttention',
+                layer_cfg=dict(  # DeformableDetrTransformerEncoderLayer
+                    self_attn_cfg=dict(  # MultiScaleDeformableAttention
                         embed_dims=256,
                         num_heads=8,
-                        num_levels=3,
+                        num_levels=3,  # 3 in DeformDETR
```
Comment (jshilong) [inline configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco-panoptic.py:64]: the comment `# 3 in DeformDETR` is not clear enough 
or just remove it

File: configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco-panoptic.py [hunk 2]
```diff
@@ -55,62 +55,45 @@
...
+                    batch_first=True),
+                cross_attn_cfg=dict(  # MultiheadAttention
                     embed_dims=256,
                     num_heads=8,
-                    attn_drop=0.0,
-                    proj_drop=0.0,
-                    dropout_layer=None,
-                    batch_first=False),
-                ffn_cfgs=dict(
+                    dropout=0.0,  # 0.1 for DETR
```
Comment (jshilong) [inline configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco-panoptic.py:89]: `0.1 for DETR` is confusing; I can not get what it means.

File: configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco.py [hunk 3]
```diff
@@ -56,6 +56,17 @@
     dict(type='PackDetInputs')
 ]
 
+test_pipeline = [
```
Comment (jshilong) [inline configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco.py:59]: Why this modification is necessary？

File: mmdet/models/layers/msdeformattn_pixel_decoder.py [hunk 4]
```diff
@@ -99,9 +77,9 @@ def __init__(self,
             input_conv_list.append(input_conv)
         self.input_convs = ModuleList(input_conv_list)
 
-        self.encoder = build_transformer_layer_sequence(encoder)
-        self.postional_encoding = build_positional_encoding(
-            positional_encoding)
+        self.encoder = Mask2FormerTransformerEncoder(**encoder)  # TODO: Notice reference_point
```
Comment (jshilong) [inline mmdet/models/layers/msdeformattn_pixel_decoder.py:80]: remember this todo

File: mmdet/models/layers/msdeformattn_pixel_decoder.py [hunk 5]
```diff
@@ -237,19 +215,14 @@ def forward(self, feats: List[Tensor]) -> Tuple[Tensor, Tensor]:
...
-            query_key_padding_mask=padding_masks,
+            key_padding_mask=padding_masks,
             spatial_shapes=spatial_shapes,
             reference_points=reference_points,
             level_start_index=level_start_index,
-            valid_radios=valid_radios)
-        # (num_total_query, batch_size, c) -> (batch_size, c, num_total_query)
-        memory = memory.permute(1, 2, 0)
+            valid_ratios=valid_radios)
+        # (batch_size, num_total_query, c) -> (batch_size, c, num_total_query)
```
Comment (jshilong) [inline mmdet/models/layers/msdeformattn_pixel_decoder.py:224]: num_total_queries

File: configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco.py (hunk 3 above)
Comment (Li-Qingyun) [reply configs/mask2former/mask2former_r50_8xb2-lsj-50e_coco.py:59]: @jshilong 
Because the origin config can not launch experiment successfully.
I understand that this config is for instance segmentation, however, the config is inherited from coco-panoptic config.

File: mmdet/models/layers/msdeformattn_pixel_decoder.py (hunk 4 above)
Comment (Li-Qingyun) [reply mmdet/models/layers/msdeformattn_pixel_decoder.py:80]: @jshilong The encoder has been customized for Mask2Former. This will be rethinked when refactoring Mask2Former.

**Comment by Li-Qingyun** [comment]
> please merge the refactor-detr and resolve the conflict

@jshilong the merging has been finished and the conflicts has been resolved.

File: mmdet/models/dense_heads/mask2former_head.py [hunk 6]
```diff
@@ -125,7 +125,8 @@ def __init__(self,
                         feat_channels, self.decoder_embed_dims, kernel_size=1))
             else:
                 self.decoder_input_projs.append(nn.Identity())
-        self.decoder_positional_encoding = MODELS.build(positional_encoding)
+        self.decoder_positional_encoding = SinePositionalEncoding(
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/mask2former_head.py:128]: I think we can keep using positional_encoding rather than positional_encoding_cfg

File: mmdet/models/layers/msdeformattn_pixel_decoder.py [hunk 7]
```diff
@@ -137,7 +116,7 @@ def __init__(self,
             feat_channels, out_channels, kernel_size=1, stride=1, padding=0)
 
         self.num_outs = num_outs
-        self.point_generator = MlvlPointGenerator(strides)
+        self.point_generator = MlvlPointGenerator(strides)  # TODO
```
Comment (ZwwWayne) [inline mmdet/models/layers/msdeformattn_pixel_decoder.py:119]: TODO of what?

File: mmdet/models/dense_heads/mask2former_head.py (hunk 6 above)
Comment (Li-Qingyun) [reply mmdet/models/dense_heads/mask2former_head.py:128]: OK, all the related parameter names have been modified.

**Review by jshilong — APPROVED** [review]
LGTM

File: mmdet/models/layers/msdeformattn_pixel_decoder.py [hunk 8]
```diff
@@ -161,9 +138,9 @@ def init_weights(self) -> None:
 
         # init_weights defined in MultiScaleDeformableAttention
         for layer in self.encoder.layers:
-            for attn in layer.attentions:
-                if isinstance(attn, MultiScaleDeformableAttention):
-                    attn.init_weights()
+            attn = layer.self_attn
```
Comment (jshilong) [inline mmdet/models/layers/msdeformattn_pixel_decoder.py:141]: I recommend using
```
 for m in self.encoder.layers.modules():
     if isinstance(attn, MultiScaleDeformableAttention):
         m.init_weights()
```

Comment (jshilong) [reply mmdet/models/layers/msdeformattn_pixel_decoder.py:141]: This can keep consistent with the original implementation even there is multiple MSDA

Comment (Li-Qingyun) [reply mmdet/models/layers/msdeformattn_pixel_decoder.py:141]: okkkkk @jshilong
